File: src/sp/it/pl/audio/playback/BalanceProperty.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package sp.it.pl.audio.playback;

import javafx.beans.property.SimpleDoubleProperty;
import static java.lang.Math.abs;
import static sp.it.pl.util.Util.clip;

/** 
 * Double property for left-right speaker volume balance, with additional 
 * methods. Makes sure the value is always within valid range.
 */
public final class BalanceProperty extends SimpleDoubleProperty {
    
    public static final double MIN = -1;
    public static final double MAX = 1;
    public static final double AVG = (MAX+MIN)/2;
    public static final double STEP = abs(MAX-MIN)/20;
    
    /** Initializes default value: 0 */
    public BalanceProperty() {
        super(AVG);
    }    
    
    /** Initializes to provided value. */
    public BalanceProperty(double v) {
        super(clip(MIN,v,MAX));
    }

    /** Sets the value. Value outside of minimal-maximal value range will be clipped. */
    @Override
    public void set(double v) {
        super.set(clip(MIN,v,MAX));
    }
    
    /** @return minimum volume value: -1 */
    public double getMin() {
        return MIN;
    }
    
    /** @return average volume value: 0 */
    public double getAverage() {
        return AVG;
    }
    
    /** @return maximum volume value: 1 */
    public double getMax() {
        return MAX;
    }
    
    
    /** Increment value by step. */
    public void left() {
        set(get()+STEP);
    }

    /** Decrement value by step. */
    public void right() {
        set(get()-STEP);
    }
}

File: src/player/main/sp/it/pl/audio/playback/BalanceProperty.kt
package sp.it.pl.audio.playback

import javafx.beans.property.SimpleDoubleProperty
import sp.it.pl.audio.playback.BalanceProperty.Companion.MAX
import sp.it.pl.audio.playback.BalanceProperty.Companion.MIN
import sp.it.util.math.clip

/** Double property for playback left-right volume balance. Value is always within valid range [MIN]-[MAX]. */
class BalanceProperty: SimpleDoubleProperty {

    /** Minimum volume value: [MIN] */
    val min get() = MIN
    /** Average volume value: [AVG] */
    val average get() = AVG
    /** Maximum volume value: [MAX] */
    val max get() = MAX

    constructor(v: Double): super(v.clip(MIN, MAX))

    /** Sets the value. Value outside of minimal-maximal value range will be clipped. */
    override fun set(v: Double) {
        super.set(v.clip(MIN, MAX))
    }

    /** Increment value by [STEP]. */
    fun leftByStep() {
        set(get()+STEP)
    }

    /** Decrement value by [STEP]. */
    fun rightByStep() {
        set(get()-STEP)
    }

    companion object {
        const val MIN = -1.0
        const val MAX = 1.0
        const val AVG = (MAX+MIN)/2
        const val STEP = 0.2
    }

}